lxd/ubuntupro/client.go

```diff
@@ -0,0 +1,261 @@
+package ubuntupro
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/canonical/lxd/lxd/fsmonitor"
+	"github.com/canonical/lxd/lxd/fsmonitor/drivers"
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/api"
+	"github.com/canonical/lxd/shared/logger"
+)
+
+// GuestAttachSetting represents Ubuntu Pro settings for guest attachment availability.
+type GuestAttachSetting string
+
+const (
+	// GuestAttachSettingOff indicates that guest attachment is turned off.
+	// - When the host has this setting turned off, devlxd requests to `GET /1.0/ubuntu-pro` should return "off" and
+	//   `POST /1.0/ubuntu-pro/token` should return a 403 Forbidden (regardless of the guest setting).
+	// - When the guest has this setting turned off (`ubuntu_pro.guest_attach`), devlxd requests to `GET /1.0/ubuntu-pro`
+	//   should return "off" and `POST /1.0/ubuntu-pro/token` should return a 403 Forbidden (regardless of the host setting).
+	GuestAttachSettingOff GuestAttachSetting = "off"
+
+	// GuestAttachSettingAvailable indicates that guest attachment is available.
+	// - When the host has this setting, devlxd requests to `GET /1.0/ubuntu-pro` should return the setting from the guest
+	//   (`ubuntu_pro.guest_attach) and `POST /1.0/ubuntu-pro/token` should retrieve a guest token via the Ubuntu Pro client.
+	// - When the guest has this setting, the pro client inside the guest will not try to retrieve a guest token at startup
+	//   (though attachment with a guest token can still be performed with `pro auto-attach`.
+	GuestAttachSettingAvailable GuestAttachSetting = "available"
+
+	// GuestAttachSettingOn indicates that guest attachment is on.
+	// - When the host has this setting, devlxd requests to `GET /1.0/ubuntu-pro` should return the setting from the guest
+	//   (`ubuntu_pro.guest_attach) and `POST /1.0/ubuntu-pro/token` should retrieve a guest token via the Ubuntu Pro client.
+	// - When the guest has this setting, the pro client inside the guest will attempt to retrieve a guest token at startup.
+	GuestAttachSettingOn GuestAttachSetting = "on"
+)
+
+// isValid returns an error if the GuestAttachSetting is not one of the pre-defined values.
+func (g GuestAttachSetting) isValid() error {
+	if !shared.ValueInSlice(g, []GuestAttachSetting{GuestAttachSettingOff, GuestAttachSettingAvailable, GuestAttachSettingOn}) {
+		return fmt.Errorf("Invalid guest auto-attach setting %q", g)
+	}
+
+	return nil
+}
+
+// Settings contains Ubuntu Pro settings relevant to LXD.
```

Comment:
Because /dev/lxd is a public API (to the applications running inside the guest) we need to ensure that the API structs for it are defined in `shared/api` like `api.DevLXDGet`.

So I think this struct should be moved into `shared/api/devlxd.go`.